### ISSUE TITLE
Generated types import the entire Apollo runtime

### ISSUE THREAD

**Issue by gsabran** [apollographql/apollo-ios#1047]
The generated types implement protocols defined in `Apollo`, such as `GraphQLQuery`. For this, they need to import `Apollo` where most of the code is about cache management and decoding (see [example](https://github.com/apollographql/apollo-ios/blob/8bcb3c630afc39046eee1fe3aec41a064fae9c6d/Sources/StarWarsAPI/API.swift#L12)).
For most apps this is fine, but in the context of a very large app where there is a focus on reducing dependencies, this is problematic and there is a need to reduce the dependencies that each generate type creates.

Ideally, things would look like (scroll right!)

```
               Apollo                  |                Consumer application
                                       |
                                       |
----------------------------------             ----------------------------------------
|          ApolloTypes           |             |             Type Module A            |
|                                |    <---     |                                      |
| protocol GraphQLQuery { ... }  |             | import ApolloTypes                   |
|________________________________|             | class MyQuery: GraphQLQuery { ... }  |      ------------------------------
                                               |______________________________________| <--- |      Feature Module A      |
                ^                                                                            |                            |
                |                                                                            | import TypeModuleA         |
                                                                                             | import Networking          | 
----------------------------------             ----------------------------------            |                            |
|            Apollo              |             |        Networking Module       |     <---   | network.send(MyQuery())    |
|                                |             |                                |            |____________________________|
| ApolloClient etc               |    <---     | Apollo                         |
| extension GraphQLQuery { ... } |             | ...                            |
|________________________________|             |________________________________|

```

where a type module would not depend on the entirety of `Apollo`

I think this is the exhaustive list of protocols required by the generated types
- `GraphQLBooleanCondition`
- `GraphQLField`
- `GraphQLFragment`
- `GraphQLFragmentSpread`
- `GraphQLID`
- `GraphQLInputValue`
- `GraphQLMap`
- `GraphQLMapConvertible`
- `GraphQLMutation`
- `GraphQLOperation`
- `GraphQLOperationType`
- `GraphQLOutputType`
- `GraphQLQuery`
- `GraphQLSelection`
- `GraphQLSelectionSet`
- `GraphQLSubscription`
- `GraphQLTypeCase`
- `GraphQLTypeCondition`
- `GraphQLVariable`
- `JSONDecodable`
- `JSONEncodable`
- `JSONDecodingError`
- `ResultMap`

**Comment by designatednerd** [comment]
I've been talking this over with @gsabran: The overall proposal here would be to create an `ApolloTypes` library that includes just the definition of these types. The main `Apollo` lib would depend on it, but the generated code would be able to depend on `ApolloTypes` only instead of the full lib. 

There's also some potential weirdness with where default protocol implementations would need to live, but that can be worked around.

I'd like to get some feedback from other people using Apollo-iOS in fairly large apps: Would you be ok with moving some APIs to an additional lib which is smaller? Basically, the question is which is more important: Keeping the number of frameworks you're using low or keeping the complexity of each framework as minimal as possible?

**Comment by valzevul** [comment]
This is a great idea. The overhead from using more frameworks is less of a problem vs exposing the caching/decoding/etc logic to every single feature module across the app.

**Comment by designatednerd** [comment]
I've started some real basic stuff on this as part of #1194 - going to keep it limited to new codegen types for now, mostly as a place to draw the line.

**Comment by calvincestari** [comment]
Hi 👋🏻 - this issue should be resolved in 1.0. 1.0 is currently in alpha releases with a full 1.0 release in H2 2022. Please take a look at the [latest alpha](https://github.com/apollographql/apollo-ios/releases) and let us know if this is still an issue for you.
